ecommerce/views_test.py

```diff
@@ -1384,6 +1384,48 @@ def test_products_viewset_detail(user_drf_client, coupon_product_ids):
     )
 
 
+def test_products_viewset_expired_programs(user_drf_client):
+    """ Test that the ProductViewSet returns only valid programs products and excludes the expired programs correctly"""
+    now = now_in_utc()
+    programs = ProgramFactory.create_batch(4)
+    runs = CourseRunFactory.create_batch(2, course__program=factory.Iterator(programs))
+    ProgramRunFactory.create_batch(
+        3,
+        program=factory.Iterator(programs[:3]),
+        end_date=factory.Iterator([None, now + timedelta(1)]),
+    )
+    ProgramRunFactory.create(program=programs[3], end_date=now - timedelta(1))
+
+    ProductVersionFactory.create_batch(
+        6, product__content_object=factory.Iterator(runs + programs)
+    )
+    response = user_drf_client.get(reverse("products_api-list"))
+    assert response.status_code == status.HTTP_200_OK
+    products = response.json()
+    expired_courseruns = CourseRun.objects.filter(enrollment_end__lt=now).values_list(
+        "id", flat=True
+    )
+    program_ids = [
+        product["latest_version"]["object_id"]
+        for product in products
+        if product["product_type"] == "program"
+    ]
+    # For all the programs in the list there should be on enrollable course run for each associated course
+    assert set(program_ids) == {programs[0].id, programs[1].id, programs[2].id}
+    # Expired program should be excluded.
+    assert programs[3].id not in program_ids
+    for program_id in program_ids:
+        program = Program.objects.get(pk=program_id)
+        count = (
+            program.courses.annotate(
+                runs=Count("courseruns", filter=~Q(courseruns__in=expired_courseruns))
+            )
+            .filter(runs=0)
+            .count()
+        )
+        assert count == 0
```

Comment:
I think this comment is misplaced because the next two lines are not checking for the enrollable courses at least (or) not related to enrollable course operation? 

I'm fine with whatever you choose for this, you can also look at my suggested changes below:

1) Move this comment above [line#1417](https://github.com/mitodl/mitxpro/pull/2379/files#diff-8eed442c6f918427c359555eb169aeeeabb4f6c4b903c7bff2598075af82627eR1417)
```suggestion
    # Expired program should be excluded.
    assert set(program_ids) == {programs[0].id, programs[1].id, programs[2].id}
    assert programs[3].id not in program_ids

    # For all the programs in the list there should be on enrollable course run for each associated course
    for program_id in program_ids:
        program = Program.objects.get(pk=program_id)
        count = (
            program.courses.annotate(
                runs=Count("courseruns", filter=~Q(courseruns__in=expired_courseruns))
            )
            .filter(runs=0)
            .count()
        )
        assert count == 0
```

2) Remove this comment and the part where we are checking the enrollable courses because i think that's already being tested in the existing test.
```suggestion
    # Expired program should be excluded.
    assert set(program_ids) == {programs[0].id, programs[1].id, programs[2].id}
    assert programs[3].id not in program_ids
```